stream_alert/alert_processor/outputs/output_base.py

```diff
@@ -243,12 +250,90 @@ def _kms_decrypt(self, data):
             response = client.decrypt(CiphertextBlob=data)
             return response['Plaintext']
         except ClientError as err:
             LOGGER.error('an error occurred during credentials decryption: %s', err.response)
 
+    @staticmethod
+    def get_formatted_output_credentials_name(service_name, descriptor):
+        """Formats the output name for this credential by combining the service
+        and the descriptor.
+
+        Args:
+            service_name (str): Service name on output class (i.e. "pagerduty", "demisto")
+            descriptor (str): Service destination (ie: slack channel, pd integration)
+
+        Returns:
+            str: Formatted credential name (ie: slack_ryandchannel)
+        """
+        cred_name = str(service_name)
+
+        # should descriptor be enforced in all rules?
+        if descriptor:
+            cred_name = '{}/{}'.format(cred_name, descriptor)
+
+        return cred_name
+
+    def get_aws_account_id(self):
+        """Returns the AWS account ID"""
+        return self._account_id
+
+    def get_secrets_bucket_name(self):
+        """Returns the bucket name of the S3 bucket to look in for encrypted credentials"""
+        return self._secrets_bucket
+
+
+class OutputDispatcher(object):
+    """OutputDispatcher is the base class to handle routing alerts to outputs
+
+    Public methods:
+        get_secrets_bucket_name: returns the name of the s3 bucket for secrets that
+            includes a unique prefix
+        get_config_service: the name of the service used by the config to store any
+            configured outputs for this service. implemented by some subclasses, but
+            subclass is not required to implement
+        format_output_config: returns a formatted version of the outputs configuration
+            that is to be written to disk
+        get_user_defined_properties: returns any properties for this output that must be
+            provided by the user. must be implemented by subclasses
+        dispatch: handles the actual sending of alerts to the configured service. must
+            be implemented by subclass
+    """
+    __metaclass__ = ABCMeta
+    __service__ = NotImplemented
+
+    # How many times it will attempt to retry something failing using backoff
+    MAX_RETRY_ATTEMPTS = 5
+
+    # _DEFAULT_REQUEST_TIMEOUT indicates how long the requests library will wait before timing
+    # out for both get and post requests. This applies to both connection and read timeouts
+    _DEFAULT_REQUEST_TIMEOUT = 3.05
+
+    def __init__(self, config):
+        self.region = REGION
+        self.config = config
+
+        self._credentials_provider = OutputCredentialsProvider(config,
+                                                               self._get_default_properties(),
+                                                               self.__service__)
+
+        self.account_id = self._credentials_provider.get_aws_account_id()
```

Comment:
since a instance property already has these values on it (self._credentials_provider), I'd suggest using read-only computed properties that 'fall through' to these. like so:


```
@property
def account_id(self):
    return self._credentials_provider.get_aws_account_id()

@property
def secrets_bucket(self):
    return self._credentials_provider.get_secrets_bucket_name()
```

Reply:
This is much better imo. It makes it so much more clear that subclasses intended to use this. I stripped out a lot of other functionality but I noticed that **one** subclass used `account_id`.

Reply:
At least for `OutputDispatcher`, what I'm trying to do is to reduce the amount of automatic inheritance.

I noticed that `OutputDispatcher` has a mix of methods; some are endemic to being an alert dispatch abstraction (e.g. `dispatch()`, `_dispatch()` or `get_user_defined_properties()` etc), but others are more like utilities (e.g. `_get_request_retry()`, or `_load_creds()` etc). For the latter, these methods are super convenient to have around for certain subclasses of `OutputDispatcher`, but are useless on the rest.

My plan is to break it up into a series of services that help `OutputDispatcher`. Then, instead of using automatic inheritance, we can use composition or mixins to assemble only the features necessary.

Reply:
👍 I like the idea of using mixins for a lot of this functionality. thanks for the info!